libxammac.dylib missing when performing dllimport call

Hello!

I'm currently attempting to add macOS support to LibVLCSharp (libvlc binding for xamarin). After some issues regarding the packaging/bundling of dylibs, I managed to copy them all in the final .app bundle. My next (and hopefully final) issue to get this working is making the `DllImport` calls work.

### Steps to Reproduce

1. Generate package from https://github.com/mfkl/libvlc-nuget/pull/5 (not on nuget yet)
2. Add it to a Xamarin.Mac project
3. Try loading the dylib at runtime and P/Invoking into it (CLI launch for interesting debug logs).

### Expected Behavior
Can perform P/Invoke calls on a dylib after loading it with `ObjCRuntime.Dlfcn.dlopen`
### Actual Behavior
Calling `dlopen` on them before 
```
 NSApplication.Init();
 NSApplication.Main(args);
```
does a return a not null pointer, but dllimport calls fail with
```
Unhandled Exception:
System.DllNotFoundException: Could not find the runtime library libxammac.dylib
  at ObjCRuntime.Runtime.LookupInternalFunction[T] (System.String name) [0x0010a] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.4.1.193/src/Xamarin.Mac/ObjCRuntime/Runtime.mac.cs:107
  at ObjCRuntime.Runtime.EnsureInitialized () [0x00030] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.4.1.193/src/Xamarin.Mac/ObjCRuntime/Runtime.mac.cs:117
  at AppKit.NSApplication.Init () [0x00016] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.4.1.193/src/Xamarin.Mac/AppKit/NSApplication.cs:56
``` 

I have a dllMap file if that matters
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <dllmap dll="libvlc" target="@executable_path/lib/libvlc.5.dylib"/>
</configuration>
```
### Environment

```
Visual Studio Community 2017 for Mac
Version 7.5.3 (build 7)
Runtime:
	Mono 5.10.1.57 (2017-12/ea8a24b1bbf) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

	Package version: 510010057

Xamarin.Mac
Version: 4.4.1.193 (Visual Studio Community)
```

I could bundle the `libxammac.dylib` in my .app bundle but... that does not feel right at all (and crashes for another reason, anyway). 

Any help or insights greatly appreciated.

> if you are creating a Xamarin.Mac App with VisualStudio for Mac you do not need to bundle libxammac.dylib at all since our tooling should do the right thing for it.

Ok, thanks for the confirmation.

> On a side note there are other projects like SkiaSharp and Urho3d that are doing the same thing you want to achieve so you might want to take a look at them too :)
> Looks like SkiaSharp uses a Xamarin.Mac binding project1 and some #if directives0 to drive the P/Invoke lib they use, neat trick.

I did have a look but I am not using objective sharpie to auto generate binding from Obj-c so I guess that's a bit different. Using #if directives for P/Invoke on iOS too btw :) https://github.com/videolan/libvlcsharp/blob/11a72ffee462d407c83b89ed94d0e4d7bf8cffd4/LibVLCSharp/Shared/Core.cs#L106

> having a sample showing your issue would be very useful, there isn't a lot we can do beyond guess from here.

My bad, here is a minimal repro: https://github.com/mfkl/LibVLC.PInvoke.Mac

> You could try setting MONO_LOG_LEVEL=debug MONO_LOG_MASK=dll to see where mono is trying to load libxammac.dylib from.

```
Martz-2:MonoBundle Martz$ mono LibVLC.PInvoke.Mac.exe
Mono: DllImport attempting to load: '/usr/lib/libSystem.dylib'.
Mono: DllImport loaded library '/usr/lib/libSystem.dylib'.
Mono: DllImport searching in: '/usr/lib/libSystem.dylib' ('/usr/lib/libSystem.dylib').
Mono: Searching for 'dlopen'.
Mono: Probing 'dlopen'.
Mono: Found as 'dlopen'.
140273887815600 # this is me printing the result of dlopen on my 2 dylibs
140273889895568 # this is me printing the result of dlopen on my 2 dylibs
Mono: DllImport attempting to load: '/usr/lib/libobjc.dylib'.
Mono: DllImport loaded library '/usr/lib/libobjc.dylib'.
Mono: DllImport searching in: '/usr/lib/libobjc.dylib' ('/usr/lib/libobjc.dylib').
Mono: Searching for 'sel_registerName'.
Mono: Probing 'sel_registerName'.
Mono: Found as 'sel_registerName'.
Mono: DllImport searching in: '/usr/lib/libobjc.dylib' ('/usr/lib/libobjc.dylib').
Mono: Searching for 'objc_getClass'.
Mono: Probing 'objc_getClass'.
Mono: Found as 'objc_getClass'.
Mono: DllImport searching in: '/usr/lib/libSystem.dylib' ('/usr/lib/libSystem.dylib').
Mono: Searching for 'dlsym'.
Mono: Probing 'dlsym'.
Mono: Found as 'dlsym'.
You are using dlopen without a full path, retrying by prepending /usr/lib

Unhandled Exception:
System.DllNotFoundException: Could not find the runtime library libxammac.dylib
  at ObjCRuntime.Runtime.LookupInternalFunction[T] (System.String name) [0x000a6] in <91a95489a34f4ae887dadc916f0228c3>:0
  at ObjCRuntime.Runtime.EnsureInitialized () [0x00030] in <91a95489a34f4ae887dadc916f0228c3>:0
  at AppKit.NSApplication.Init () [0x00016] in <91a95489a34f4ae887dadc916f0228c3>:0
  at LibVLC.PInvoke.Mac.MainClass.Main (System.String[] args) [0x00049] in <10a0767b1d2d469b880d78cded2947b0>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: Could not find the runtime library libxammac.dylib
  at ObjCRuntime.Runtime.LookupInternalFunction[T] (System.String name) [0x000a6] in <91a95489a34f4ae887dadc916f0228c3>:0
  at ObjCRuntime.Runtime.EnsureInitialized () [0x00030] in <91a95489a34f4ae887dadc916f0228c3>:0
  at AppKit.NSApplication.Init () [0x00016] in <91a95489a34f4ae887dadc916f0228c3>:0
  at LibVLC.PInvoke.Mac.MainClass.Main (System.String[] args) [0x00049] in <10a0767b1d2d469b880d78cded2947b0>:0
```

I'd assume that once loaded with `ObjCRuntime.Dlfcn.dlopen`, the librairies are ready for P/Invoke but I guess that's a wrong assumption (as P/Invoke does a dlopen anyway).

I'm seeing this when I try to build. Are all of your changes landed there?

```    xcrun -sdk macosx install_name_tool -id @executable_path/../MonoBundle/plugins/"" ""
    EXEC : error : /Applications/Xcode94.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: can't open file:  (No such file or directory)
    /Users/donblas/Programming/Projects/LibVLC.PInvoke.Mac/LibVLC.PInvoke.Mac/LibVLC.PInvoke.Mac.csproj(121,5): error MSB3073: The command "xcrun -sdk macosx install_name_tool -id @executable_path/../MonoBundle/plugins/"" """ exited with code 1.```

--- a/docs/website/generator-errors.md
+++ b/docs/website/generator-errors.md
@@ -216,6 +216,14 @@ This usually indicates a bug in Xamarin.iOS/Xamarin.Mac; please [file a bug repo
 
 ### <a name='BI1117'/>BI1117: The member '*' is decorated with [Static] and its container class * is decorated with [Category] this leads to hard to use code. Please inline * into * class.
 
+### <a name='BI1118'/>[NullAllowed] should not be used on methods, like '*', but only on properties, parameters and return values.
+
+The `[NullAllowed]` attribute should not be allowed on methods but it could break existing binding projects.
+
+Historically it was used on property setters. However using the attribute on _other_ methods can be misleading, e.g. should it apply to all parameters, the return value... and its presence/action can be misinterpreted in code reviews leading to binding bugs.
+
+To fix this warning use the `[NullAllowed]` attribute onnly on parameters, properties or return values.
+
 <!-- 2xxx: reserved -->
 <!-- 3xxx: reserved -->
 <!-- 4xxx: reserved -->

--- a/docs/website/generator-errors.md
+++ b/docs/website/generator-errors.md
@@ -222,7 +222,7 @@ The `[NullAllowed]` attribute should not be allowed on methods but it could brea
 
 Historically it was used on property setters. However using the attribute on _other_ methods can be misleading, e.g. should it apply to all parameters, the return value... and its presence/action can be misinterpreted in code reviews leading to binding bugs.
 
-To fix this warning use the `[NullAllowed]` attribute onnly on parameters, properties or return values.
+To fix this warning use the `[NullAllowed]` attribute only on parameters, properties or return values.
 
 <!-- 2xxx: reserved -->
 <!-- 3xxx: reserved -->

--- a/src/generator.cs
+++ b/src/generator.cs
@@ -4096,6 +4096,9 @@ public partial class Generator : IMemberGatherer {
 		if (null_allowed_override)
 			return;
 
+		if (AttributeManager.HasAttribute<NullAllowedAttribute> (mi))
+			ErrorHelper.Show (new BindingException (1118, false, $"[NullAllowed] should not be used on methods, like '{mi}', but only on properties, parameters and return values."));
+
 		foreach (var pi in mi.GetParameters ()) {
 			var needs_null_check = ParameterNeedsNullCheck (pi, mi, propInfo);
 			if (!needs_null_check)
@@ -4808,7 +4811,10 @@ public partial class Generator : IMemberGatherer {
 		if (pi.CanWrite){
 			var setter = pi.GetSetMethod ();
 			var ba = GetBindAttribute (setter);
-			bool null_allowed = AttributeManager.HasAttribute<NullAllowedAttribute> (pi) || AttributeManager.HasAttribute<NullAllowedAttribute> (setter);
+			bool null_allowed = AttributeManager.HasAttribute<NullAllowedAttribute> (setter);
+			if (null_allowed)
+				ErrorHelper.Show (new BindingException (1118, false, $"[NullAllowed] should not be used on methods, like '{setter}', but only on properties, parameters and return values."));
+			null_allowed |= AttributeManager.HasAttribute<NullAllowedAttribute> (pi);
 			var not_implemented_attr = AttributeManager.GetCustomAttribute<NotImplementedAttribute> (setter);
 			string sel;
 

--- a/tests/generator/BGenTests.cs
+++ b/tests/generator/BGenTests.cs
@@ -564,6 +564,28 @@ namespace GeneratorTests
 		[Test]
 		public void GHIssue5444 () => BuildFile (Profile.iOS, "ghissue5444.cs");
 
+		[Test]
+		public void GH5416_method ()
+		{
+			var bgen = new BGenTool ();
+			bgen.Profile = Profile.iOS;
+			bgen.AddTestApiDefinition ("ghissue5416b.cs");
+			bgen.CreateTemporaryBinding ();
+			bgen.AssertExecute ("build");
+			bgen.AssertWarning (1118, "[NullAllowed] should not be used on methods, like 'NSString Method(Foundation.NSDate, Foundation.NSObject)', but only on properties, parameters and return values.");
+		}
+
+		[Test]
+		public void GH5416_setter ()
+		{
+			var bgen = new BGenTool ();
+			bgen.Profile = Profile.iOS;
+			bgen.AddTestApiDefinition ("ghissue5416a.cs");
+			bgen.CreateTemporaryBinding ();
+			bgen.AssertExecute ("build");
+			bgen.AssertWarning (1118, "[NullAllowed] should not be used on methods, like 'Void set_Setter(Foundation.NSString)', but only on properties, parameters and return values.");
+		}
+
 		BGenTool BuildFile (Profile profile, params string [] filenames)
 		{
 			return BuildFile (profile, true, false, filenames);

--- a/tests/generator/ghissue5416a.cs
+++ b/tests/generator/ghissue5416a.cs
@@ -1,0 +1,11 @@
+using Foundation;
+using ObjCRuntime;
+
+namespace GHIssue5416 {
+
+	[BaseType (typeof (NSObject))]
+	interface NullAllowedWarning {
+		[Export ("setter", ArgumentSemantic.Assign)]
+		NSString Setter { get; [NullAllowed] set; }
+	}
+}

--- a/tests/generator/ghissue5416b.cs
+++ b/tests/generator/ghissue5416b.cs
@@ -1,0 +1,12 @@
+using Foundation;
+using ObjCRuntime;
+
+namespace GHIssue5416 {
+
+	[BaseType (typeof (NSObject))]
+	interface NullAllowedWarning {
+		[Export ("methodWhen:data:", ArgumentSemantic.Assign)]
+		[NullAllowed]
+		NSString Method (NSDate d, NSObject o);
+	}
+}

